scapy/contrib/mqttsn.py

```diff
@@ -0,0 +1,468 @@
+# This file is part of Scapy
+# See http://www.secdev.org/projects/scapy for more information
+# Copyright (C) 2019 Freie Universitaet Berlin
+# This program is published under GPLv2 license
+#
+# Specification:
+#   http://www.mqtt.org/new/wp-content/uploads/2009/06/MQTT-SN_spec_v1.2.pdf
+
+# scapy.contrib.description = MQTT for Sensor Networks (MQTT-SN)
+# scapy.contrib.status = loads
+
+from scapy.packet import Packet, bind_layers, bind_bottom_up
+from scapy.fields import BitField, BitEnumField, ByteField, ByteEnumField, \
+    ConditionalField, FieldLenField, ShortField, StrFixedLenField, \
+    StrLenField, XByteEnumField
+from scapy.layers.inet import UDP
+from scapy.error import Scapy_Exception
+from scapy.compat import chb, orb
+from scapy.volatile import RandNum
+
+
+# Constants
+ADVERTISE = 0x00
+SEARCHGW = 0x01
+GWINFO = 0x02
+CONNECT = 0x04
+CONNACK = 0x05
+WILLTOPICREQ = 0x06
+WILLTOPIC = 0x07
+WILLMSGREQ = 0x08
+WILLMSG = 0x09
+REGISTER = 0x0a
+REGACK = 0x0b
+PUBLISH = 0x0c
+PUBACK = 0x0d
+PUBCOMP = 0x0e
+PUBREC = 0x0f
+PUBREL = 0x10
+SUBSCRIBE = 0x12
+SUBACK = 0x13
+UNSUBSCRIBE = 0x14
+UNSUBACK = 0x15
+PINGREQ = 0x16
+PINGRESP = 0x17
+DISCONNECT = 0x18
+WILLTOPICUPD = 0x1a
+WILLTOPICRESP = 0x1b
+WILLMSGUPD = 0x1c
+WILLMSGRESP = 0x1d
+ENCAPS_MSG = 0xfe
+
+QOS_0 = 0b00
+QOS_1 = 0b01
+QOS_2 = 0b10
+QOS_NEG1 = 0b11
+
+TID_NORMAL = 0b00
+TID_PREDEF = 0b01
+TID_SHORT = 0b10
+TID_RESVD = 0b11
+
+
+ACCEPTED = 0x00
+REJ_CONJ = 0x01
+REJ_TID = 0x02
+REJ_NOTSUP = 0x03
+
+
+# Custom fields
+class VariableFieldLenField(FieldLenField):
+    """
+    MQTT-SN length field either has 1 byte for values [0x02, 0xff] or 3 bytes
+    for values [0x0100, 0xffff]. If the first byte is 0x01 the length value
+    comes in network byte-order in the next 2 bytes. MQTT-SN packets are at
+    least 2 bytes long (length field + type field).
+    """
+    def addfield(self, pkt, s, val):
+        val = self.i2m(pkt, val)
+        if (val < 2) or (val > 0xffff):
+            raise Scapy_Exception("%s: invalid length field value" %
+                                  self.__class__.__name__)
+        elif val > 0xff:
+            return s + b"".join([chb(0x01), chb(val >> 8), chb(val & 0xff)])
```

Comment:
something like `b"\x01" + struct.pack("!H", val)` would probably be more appropriate

Reply:
Too used to C programming ^^